src/main/java/com/learning/mfscreener/service/PortfolioService.java

```diff
@@ -115,54 +124,66 @@ public String upload(MultipartFile multipartFile) throws IOException {
         List<UserTransactionDetailsEntity> userTransactionDetailsEntityList =
                 this.userTransactionDetailsService.findAllTransactionsByEmailAndName(email, name);
-
-        UserCASDetailsEntity userCASDetailsEntity = casDetailsEntityRepository.findByInvestorEmailAndName(email, name);
+        UserCASDetailsEntity userCASDetailsEntity = userCASDetailsService.findByInvestorEmailAndName(email, name);
 
         if (userTransactionDTOListCount == userTransactionDetailsEntityList.size()) {
             LOGGER.info("No new transactions are added");
             return null;
         }
 
+        return processFoliosAndTransactions(
+                email,
+                name,
+                casDTO,
+                userCASDetailsEntity,
+                userTransactionDTOListCount,
+                userTransactionDetailsEntityList);
+    }
+
+    UploadResponseHolder processFoliosAndTransactions(
+            String email,
+            String name,
+            CasDTO casDTO,
+            UserCASDetailsEntity userCASDetailsEntity,
+            long userTransactionDTOListCount,
+            List<UserTransactionDetailsEntity> userTransactionDetailsEntityList) {
         AtomicInteger folioCounter = new AtomicInteger();
         AtomicInteger transactionsCounter = new AtomicInteger();
 
-        // Verify if new folios are added
-        List<String> existingFolios = userCASDetailsEntity.getFolioEntities().stream()
-                .map(UserFolioDetailsEntity::getFolio)
-                .toList();
+        processNewFolios(casDTO.folios(), userCASDetailsEntity, folioCounter, transactionsCounter);
+        updateSchemesAndTransactions(
+                email,
+                name,
+                casDTO,
+                userCASDetailsEntity,
+                userTransactionDTOListCount,
+                userTransactionDetailsEntityList,
+                folioCounter,
+                transactionsCounter);
 
-        folioDTOList.forEach(userFolioDTO -> {
-            String folio = userFolioDTO.folio();
-            if (!existingFolios.contains(folio)) {
-                LOGGER.info("New folio: {} created that is not present in the database", folio);
-                userCASDetailsEntity.addFolioEntity(
-                        casDetailsMapper.mapUserFolioDTOToUserFolioDetailsEntity(userFolioDTO));
-                folioCounter.incrementAndGet();
-                int newTransactions = userFolioDTO.schemes().stream()
-                        .map(UserSchemeDTO::transactions)
-                        .flatMap(List::stream)
-                        .toList()
-                        .size();
-                transactionsCounter.addAndGet(newTransactions);
-            }
-        });
+        return new UploadResponseHolder(userCASDetailsEntity, folioCounter.get(), transactionsCounter.get());
+    }
 
+    void updateSchemesAndTransactions(
```

Comment:
@coderabbitai, suggest code changes without breaking functionality to reduce cognitive complexity.

Reply:
@coderabbitai, in above review please suggest how to reduce cognitive complexity for updateSchemesAndTransactions method without losing functionality.